File: resolution/parser.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <tuple>
#include <vector>
#include <cstdlib>
#include <unistd.h>
#include <filesystem>
#include <cctype>
#include <chrono>

using namespace std;

void make_domain() {
    string std_domain = R"((define
    (domain domain-lightsup)
    (:requirements :strips :typing :negative-preconditions)
    (:types
        line col
    )
    
    (:predicates
        (is-on ?l - line ?c - col)
        (is-visible ?l1 - line ?c1 - col ?l2 - line ?c2 - col)
        (is-numbered-adjacent ?l1 - line ?c1 - col ?l2 - line ?c2 - col)
        (counter-at ?l - line ?c - col ?cnt - counter)
    )

    (:constants i0 i1 i2 i3 i4 - counter)
    
    (:action bulb
    
        :parameters(?l - line ?c - col)
        :precondition (not (is-on ?l ?c))
        
        :effect (and

            (is-on ?l ?c)
            
            (forall (?c2 - col) (and
                ; Ativando as células da mesma linha
                (when (or 
                        (is-visible ?l ?c ?l ?c2)
                        (is-visible ?l ?c2 ?l ?c)
                    )
                        (is-on ?l ?c2)
                )

                (when (is-numbered-adjacent ?l ?c ?l ?c2) (and
                    (when (counter-at ?l ?c2 i0) (and (not (counter-at ?l ?c2 i0)) (counter-at ?l ?c2 i1)))
                    (when (counter-at ?l ?c2 i1) (and (not (counter-at ?l ?c2 i1)) (counter-at ?l ?c2 i2)))
                    (when (counter-at ?l ?c2 i2) (and (not (counter-at ?l ?c2 i2)) (counter-at ?l ?c2 i3)))
                    (when (counter-at ?l ?c2 i3) (and (not (counter-at ?l ?c2 i3)) (counter-at ?l ?c2 i4)))
                ))

            ))
            
            (forall (?l2 - line) (and
                ; Ativando as células da mesma coluna
                (when (or 
                        (is-visible ?l ?c ?l2 ?c)
                        (is-visible ?l2 ?c ?l ?c)
                    )
                        (is-on ?l2 ?c)
                )

                (when (is-numbered-adjacent ?l ?c ?l2 ?c) (and
                    (when (counter-at ?l2 ?c i0) (and (not (counter-at ?l2 ?c i0)) (counter-at ?l2 ?c i1)))
                    (when (counter-at ?l2 ?c i1) (and (not (counter-at ?l2 ?c i1)) (counter-at ?l2 ?c i2)))
                    (when (counter-at ?l2 ?c i2) (and (not (counter-at ?l2 ?c i2)) (counter-at ?l2 ?c i3)))
                    (when (counter-at ?l2 ?c i3) (and (not (counter-at ?l2 ?c i3)) (counter-at ?l2 ?c i4)))
                ))
            ))
        )
    )
))";

    ofstream domain("domain.pddl");
    domain << std_domain;
    domain.close();

}

vector<vector<char>> make_board() {

    vector<vector<char>> board; // Vetor de vetores para o board
    string line;

    // Ler line por line do input
    while (getline(cin, line)) {

        vector<char> line_vetor;

        // Processar cada caracter da line
        for (char c : line) {
            line_vetor.push_back(c);
        }

        board.push_back(line_vetor);
    }

    return board;

}

void make_problem (vector<vector<char>> board){


    // ---------------- STRING PADRÃO ----------------------------

    string std_problem = R"((define
        (problem problem-lightsup)
        (:domain domain-lightsup)
        (:objects )
        (:init )
        (:goal (and ))))";


    string OBJ = ":objects ";
    string INIT = ":init ";
    string GOAL = "(and ";

    // -----------------------------------------------------------
    // ---------------- MANIPULAÇÃO DA STRING --------------------
    // -----------------------------------------------------------

    string objects = "", init = "", goal = "";
    char c1, c2;

    // Loop para ditar visibilidade
    for (int i = 0; i < board.size(); i++) {

        string str_i = to_string(i);

        objects += "l" + str_i + " ";

        for (int j = 0; j < board[0].size(); j++) {

            string str_j = to_string(j);

            c1 = board[i][j];

            if (c1 == '-') {

                // Loopando mesma linha
                for (int jj = j + 1; jj < board[i].size(); jj++) {

                    c2 = board[i][jj];

                    if (c2 == '#' || isdigit(c2)) {
                        break;
                    }

                    init += "(is-visible l" + str_i + " c" + str_j + " l" + str_i + " c" + to_string(jj) + ")\n";
                }

                // Loopando mesma coluna
                for (int ii = i + 1; ii < board.size(); ii++) {

                    c2 = board[ii][j];

                    if (c2 == '#' || isdigit(c2)) {
                        break;
                    }

                    init += "(is-visible l" + str_i + " c" + str_j + " l" + to_string(ii) + " c" + str_j + ")\n";
                }

                goal += "(is-on l" + str_i + " c" + str_j +")\n";
            
            } else if (isdigit(c1)) {

                init += "(counter-at l" + str_i + " c" + str_j + " i0)\n";
                goal += "(counter-at l" + str_i + " c" + str_j + " i" + c1 + ")\n";

                // Verificando conexões com quadrados brancos
                // Baixo
                if (i + 1 < board.size()) {

                    c2 = board[i + 1][j];

                    if (c2 == '-') {
                        init += "(is-numbered-adjacent l" + to_string(i + 1) + " c" + str_j + " l" + str_i + " c" + str_j + ")\n";
                    }
                }

                // Cima
                if (i - 1 > -1) {

                    c2 = board[i - 1][j];

                    if (c2 == '-') {
                        init += "(is-numbered-adjacent l" + to_string(i - 1) + " c" + str_j + " l" + str_i + " c" + str_j + ")\n";
                    }
                }

                // Direita
                if (j + 1 < board[0].size()) {

                    c2 = board[i][j + 1];

                    if (c2 == '-') {
                        init += "(is-numbered-adjacent l" + str_i + " c" + to_string(j + 1) + " l" + str_i + " c" + str_j + ")\n";
                    }
                }

                // Esquerda
                if (j - 1 > -1) {

                    c2 = board[i][j - 1];
                    if (c2 == '-') {
                        init += "(is-numbered-adjacent l" + str_i + " c" + to_string(j - 1) + " l" + str_i + " c" + str_j + ")\n";
                    }
                }
            }

            // Colocando as colunas
            if (i == board.size() - 1) {
                if (j == 0) {
                    objects += "- line \n";
                }

                objects += "c" + str_j + " ";
            }
        }
    }

    objects += "- col\n";

    std_problem.replace(std_problem.find(GOAL), GOAL.size(), GOAL + goal);
    std_problem.replace(std_problem.find(INIT), INIT.size(), INIT + init);
    std_problem.replace(std_problem.find(OBJ), OBJ.size(), OBJ + objects);

    // -----------------------------------------------------------
    // ---------------- CRIANDO O ARQUIVO ------------------------
    // -----------------------------------------------------------

    ofstream problem("problem.pddl");
    problem << std_problem;
    problem.close();

}

void cat_plan() {

    if (! filesystem::exists("plan.txt")) {
        exit(120);
    }

    system("cat plan.txt | grep 'STEP' | tr -d '[:space:]' | sed -E 's/STEP[0-9]+://g; s/\\(l/ /g; s/,c/ /g; s/bu/(bu/g; s/\\)/\\);/g; s/.$//g'");
}

int main() {

    auto start = chrono::high_resolution_clock::now();

    // string exec_command = "../planners/downward-fdss23/fast-downward.py --alias lama-first --overall-time-limit 29s --plan-file plan.txt domain.pddl problem.pddl  > /dev/null";
    string exec_command = "../planners/madagascar/Mp -o plan.txt -S 1 domain.pddl problem.pddl > /dev/null";
    make_domain();
    make_problem(make_board());
    system(exec_command.c_str());
    cat_plan();

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;
    cout << "Tempo de execução: " << duration.count() << " segundos" << '\n';

    return 0;
}

File: resolution/parser.py
import sys
import os

DOMAIN_STR = """(define
    (domain domain-lightsup)
    (:requirements :strips :adl :typing :negative-preconditions)
    (:types
        position counter
    )

    (:constants i0 i1 i2 i3 i4 - counter)
    
    (:predicates
        (cell-lightened ?p - position)
        (counter-at ?p - position ?c - counter)
    )
actions_here
)"""

PROBLEM_STR = """(define
        (problem problem-lightsup)
        (:domain domain-lightsup)
        (:objects )
        (:init )
        (:goal (and )))"""


def get_input():

    vet = []

    for index, line in enumerate(sys.stdin):

        vet.append([])

        for char in line:
            if char != '\n':
                vet[index].append(char)

    return vet


def get_name(symbol: str, line: str, col: str) -> str:
    if symbol.isdigit():
        name = f"p_{line}_{col}_{symbol}"

    elif symbol == "#" or symbol == '-':
        name = f"p_{line}_{col}"

    else:
        name = None

    return name


def add_in_dict(my_dict: dict, position: str, block: str = "", numbered: str = ""):

    if position not in my_dict:
        my_dict[position] = [set(), set()]
        
    if block:
        my_dict[position][0].add(block)
    if numbered:
        my_dict[position][1].add(numbered)


def add_cell_visibility(main_dict: dict, board: list):

    # Loop em linhas
    for i in range(len(board)):
        for j in range(len(board[0])):

            current_cell = board[i][j]

            if current_cell == '-':
                current_cell_name = get_name(current_cell, i, j)
                add_in_dict(main_dict, current_cell_name)

                # Loop para esquerda
                l = j - 1
                while l >= 0:
                    previous_cell = board[i][l]

                    if previous_cell == '-':
                        previous_cell_name = get_name(previous_cell, i, l)
                        add_in_dict(main_dict, current_cell_name, previous_cell_name)
                    else:
                        break

                    l -= 1

                # Loop para direita
                r = j + 1
                while r < len(board[0]):
                    next_cell = board[i][r]

                    if next_cell == '-':
                        next_cell_name = get_name(next_cell, i, r)
                        add_in_dict(main_dict, current_cell_name, next_cell_name)
                    else:
                        break

                    r += 1

                # Loop para cima
                u = i - 1
                while u >= 0:
                    previous_cell = board[u][j]

                    if previous_cell == '-':
                        previous_cell_name = get_name(previous_cell, u, j)
                        add_in_dict(main_dict, current_cell_name, previous_cell_name)
                    else:
                        break

                    u -= 1

                # Loop para baixo
                d = i + 1
                while d < len(board):
                    next_cell = board[d][j]

                    if next_cell == '-':
                        next_cell_name = get_name(next_cell, d, j)
                        add_in_dict(main_dict, current_cell_name, next_cell_name)
                    else:
                        break

                    d += 1
    

def add_numered(main_dict: dict, board: list):

    for i in range(len(board)):
        for j in range(len(board[0])):

            current_cell = board[i][j]

            if current_cell == '-':

                # Verificando se há células numeradas (cima, baixo, direita, esquerda)
                # Cima
                if i - 1 > -1:
                    if board[i - 1][j].isdigit():
                        numbered_cell_name = get_name(board[i - 1][j], i - 1, j)
                        current_cell_name = get_name(current_cell, i, j)

                        add_in_dict(main_dict, current_cell_name, numbered=numbered_cell_name)

                # Baixo
                if i + 1 < len(board):
                    if board[i + 1][j].isdigit():
                        numbered_cell_name = get_name(board[i + 1][j], i + 1, j)
                        current_cell_name = get_name(current_cell, i, j)

                        add_in_dict(main_dict, current_cell_name, numbered=numbered_cell_name)

                # Esquerda
                if j - 1 > -1:
                    if board[i][j - 1].isdigit():
                        numbered_cell_name = get_name(board[i][j - 1], i, j - 1)
                        current_cell_name = get_name(current_cell, i, j)

                        add_in_dict(main_dict, current_cell_name, numbered=numbered_cell_name)

                # Direita
                if j + 1 < len(board[0]):
                    if board[i][j + 1].isdigit():
                        numbered_cell_name = get_name(board[i][j + 1], i, j + 1)
                        current_cell_name = get_name(current_cell, i, j)

                        add_in_dict(main_dict, current_cell_name, numbered=numbered_cell_name)


def store_positions_info(main_dict: dict, board: list):

    add_cell_visibility(main_dict, board)
    add_numered(main_dict, board)


def update_domain_str(main_dict: dict):

    actions = ""

    for key, value in main_dict.items():

        preconditions = f"(not (cell-lightened {key}))"
        
        effect = " ".join(f"(cell-lightened {cell})" for cell in value[0])
        effect +=  f" (cell-lightened {key})"
        effect_part_two = ""

        for numbered_cell in value[1]:
            effect_part_two += f"""(when (counter-at {numbered_cell} i0) (and (not (counter-at {numbered_cell} i0)) (counter-at {numbered_cell} i1)))
            (when (counter-at {numbered_cell} i1) (and (not (counter-at {numbered_cell} i1)) (counter-at {numbered_cell} i2)))
            (when (counter-at {numbered_cell} i2) (and (not (counter-at {numbered_cell} i2)) (counter-at {numbered_cell} i3)))
            (when (counter-at {numbered_cell} i3) (and (not (counter-at {numbered_cell} i3)) (counter-at {numbered_cell} i4)))\n\t\t"""

        actions += f"""(:action {key}

        :parameters()
        :precondition (and {preconditions})
        :effect(and {effect + effect_part_two})
)\n\n"""

    new_domain_str =  DOMAIN_STR.replace("actions_here", actions)
    return new_domain_str


def create_domain_file(main_dict: dict):

    domain_str = update_domain_str(main_dict)

    with open("domain.pddl", "w") as domain_file:
        domain_file.write(domain_str)


def update_problem_str(main_dict: dict):
    positions = ''
    init = ''
    goal = ''

    for key, value in main_dict.items():
        
        positions += key + " "

        goal += f"(cell-lightened {key})\n"

        for numbered_cell in value[1]:
            if numbered_cell not in init:
                init += f"(counter-at {numbered_cell} i0)\n"
                positions += numbered_cell + " "
            
            if numbered_cell not in goal:
                goal += f"(counter-at {numbered_cell} i{numbered_cell[-1]})\n"

    positions += '- position'

    new_problem_str = PROBLEM_STR.replace("objects ", f"objects {positions}")
    new_problem_str = new_problem_str.replace("init ", f"init  {init}")
    new_problem_str = new_problem_str.replace("and ", f"and {goal}")

    return new_problem_str


def create_problem_file(main_dict: dict):
    
    problem_str = update_problem_str(main_dict)

    with open("problem.pddl", "w") as domain_file:
        domain_file.write(problem_str)


def cat_plan():
    
    if not os.path.exists("plan.txt"):
        exit(120)
    
    os.system("cat plan.txt | grep 'STEP' | tr -d '[:space:]' | sed -E 's/STEP[0-9]+://g; s/_/ /g; s/p/\\(bulb/g; s/\\(\\)\\(/\\);\\(/g; s/\\(\\)/\\)/g'")

if __name__ == "__main__":

    main_dict = {}
    board = get_input()

    # /tmp/dir/software/planners/downward/fast-downward.py --plan-file plan.txt --alias seq-opt-fdss-2023 --overall-time-limit 179s domain.pddl problem.pddl > /dev/null"
    # planners/madagascar/Mp -o plan.txt -S 1 domain.pddl problem.pddl"
    execution_str = "../planners/madagascar/Mp -o plan.txt -S 1 domain.pddl problem.pddl > /dev/null"

    store_positions_info(main_dict, board)
    create_domain_file(main_dict)
    create_problem_file(main_dict)
    os.system(execution_str)
    cat_plan()